### ISSUE TITLE
REST API and ipfs-cluster-ctl should be able to talk over libp2p

### ISSUE THREAD

**Issue by hsanjuan** [ipfs/ipfs-cluster#305]
This gives secure transport out of the box, without the need to setup https. But there are many considerations:

* Might need an extra libp2p host on the rest api side, and one on the ctl side. 
* Not sure what to do about authorization. If the main libp2p host is re-used, allowing anyone to use the rest api would imply revealing the swarm secret too (bad).
* Not sure if this could be solved by registering RPC methods from the REST API and using them directly, thus, no longer an http rest api, but providing another libp2p-based api.
* Alternatively, we can pass http with libp2p. This problably comes handy: https://github.com/hsanjuan/go-libp2p-http, but should be re-written using https://github.com/hsanjuan/go-libp2p-gostream (would basic auth still work with this? maybe yes, which is nice).
* HTTP api should stay as an option (even maybe running both apis in parallel)
* REST client should support both things (cluster-ctl uses it, and cluster-ctl should support both things too)
* This will require configuration file for cluster-ctl (`ctl.conf`)
* Does it make sense to make a rest/api wrapper that does it (as a separate component, like a libp2p-to-http proxy) or rather to develop it in the rest/api component directly?

In any case, HTTP APIs are very important for integrations with other software, so providing via libp2p should be an alternative only, and the user should choose which one is on (or both).

Todo: actually figure out the best approach.

**Comment by hsanjuan** [comment]
This will happen on #349 .

tl;dr: 

* the restapi component will be able to listen and serve http via a provided libp2p host OR a configured one. By default, we will provide the cluster main host for the task.
* HTTP endpoint can be disabled. But they can run in parallel.
* This should be flexible enough. Default usecase covers using `ipfs-cluster-ctl` to communicate securely with a cluster to which the secret is known.
